yjit/src/stats.rs

```diff
@@ -231,6 +298,57 @@ fn rb_yjit_gen_stats_dict() -> VALUE {
     hash
 }
 
+/// Record the backtrace when a YJIT exit occurs. This functionality requires
+/// that the stats feature is enabled as well as the --yjit-stats and --yjit-trace-exits
+/// options are set.
+///
+/// This function will fill two Vec's in CodegenGlobals to record the raw samples
+/// and line samples. Their length should be the same, however the data stored in
+/// them is different.
+#[no_mangle]
+pub extern "C" fn rb_yjit_record_exit_stack(exit_pc: *const VALUE)
+{
+    // Return if the following are not enabled: yjit, stats fature, --yjit-stats, and
+    // --yjit-trace-exits
+    if !yjit_enabled_p() && !cfg!(feature = "stats") && !get_option!(gen_stats) && !get_option!(gen_trace_exits) {
+        return;
+    }
+
+    // Get the opcode from the encoded insn handler at this PC
+    let insn = unsafe { rb_vm_insn_addr2opcode((*exit_pc).as_ptr()) };
+    const BUFF_LEN: usize = 2048;
+    let mut frames_buffer = [VALUE(0 as usize); BUFF_LEN];
+    let mut lines_buffer = [0; BUFF_LEN];
+    let num = unsafe { rb_profile_frames(0, BUFF_LEN as i32, frames_buffer.as_mut_ptr(), lines_buffer.as_mut_ptr()) };
+    let mut i = num - 1;
```

Comment:
What does rb_profile_frames do, and where does the 2048 come from?

Reply:
`rb_profile_frames` is [the CRuby API for getting a "profiling" stack trace](https://github.com/ruby/ruby/blob/8b4d2a50148b9944dd2d1d2d6ebe7b09874ddfdd/vm_backtrace.c#L1547).  2048 came from the [original stackprof implementation](https://github.com/tmm1/stackprof/blob/df24b85953bb45d3abff58d9c82169a3003a60f1/ext/stackprof/stackprof.c#L20), but I don't think there's any real reason for it.  `rb_profile_frames` doesn't allocate anything, only write in to a buffer you provide.  2048 seemed to be working well for collecting stats on stackprof, so it seemed natural to use here.